File: dump_excel.py
import zipfile
import xml.etree.ElementTree as ET
import os

try:
    excel_path = os.path.join('SEO Files', 'issues_overview_report.xlsx')
    if not os.path.exists(excel_path):
        print(f"Error: {excel_path} not found!")
        exit(1)

    with zipfile.ZipFile(excel_path) as z:
        # 1. Read shared strings
        shared_strings = []
        if 'xl/sharedStrings.xml' in z.namelist():
            ss_data = z.read('xl/sharedStrings.xml')
            ss_root = ET.fromstring(ss_data)
            ns = {'ns': 'http://schemas.openxmlformats.org/spreadsheetml/2006/main'}
            for si in ss_root.findall('ns:si', ns):
                parts = []
                for t in si.findall('.//ns:t', ns):
                    if t.text:
                        parts.append(t.text)
                shared_strings.append(''.join(parts))

        # 2. Read sheet1
        sheet_files = [name for name in z.namelist() if name.startswith('xl/worksheets/')]
        if not sheet_files:
            print("No sheet files found!")
            exit(1)
            
        sheet_data = z.read(sheet_files[0])
        sheet_root = ET.fromstring(sheet_data)
        
        ns = {'ns': 'http://schemas.openxmlformats.org/spreadsheetml/2006/main'}
        rows = []
        
        for row in sheet_root.findall('.//ns:row', ns):
            row_num = int(row.get('r'))
            cells = {}
            for c in row.findall('ns:c', ns):
                r = c.get('r')
                col_letter = ''.join([char for char in r if char.isalpha()])
                t = c.get('t')
                
                val_text = ''
                if t == 'inlineStr':
                    is_el = c.find('ns:is', ns)
                    if is_el is not None:
                        t_el = is_el.find('ns:t', ns)
                        if t_el is not None:
                            val_text = t_el.text or ''
                else:
                    v = c.find('ns:v', ns)
                    if v is not None:
                        val_text = v.text or ''
                        
                    if t == 's' and val_text:
                        str_idx = int(val_text)
                        if str_idx < len(shared_strings):
                            val_text = shared_strings[str_idx]
                            
                cells[col_letter] = val_text
            rows.append((row_num, cells))

    # Build a clean Markdown table of the contents
    md = "# Screaming Frog Excel Audit — Full Issue Details\n\n"
    md += "| Issue Name | Issue Type | Priority | URLs | % of Total | Description | How To Fix | Help URL |\n"
    md += "|---|---|---|---|---|---|---|---|\n"

    # Sort rows by row number
    rows.sort(key=lambda x: x[0])

    for row_num, cells in rows:
        if row_num == 1:
            continue # Header row
        name = cells.get('A', '')
        itype = cells.get('B', '')
        prio = cells.get('C', '')
        urls = cells.get('D', '')
        pct = cells.get('E', '')
        desc = cells.get('F', '')
        fix = cells.get('G', '')
        help_url = cells.get('H', 'N/A') or 'N/A'
        
        # Clean up any newlines or formatting
        name = name.replace('\n', ' ').strip()
        itype = itype.replace('\n', ' ').strip()
        prio = prio.replace('\n', ' ').strip()
        desc = desc.replace('\n', ' ').strip()
        fix = fix.replace('\n', ' ').strip()
        
        md += f"| **{name}** | {itype} | {prio} | {urls} | {pct}% | {desc} | {fix} | {help_url} |\n"

    output_path = os.path.join('SEO Files', 'excel_issues_dump.md')
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(md)

    print("SUCCESS: Generated excel_issues_dump.md")
except Exception as e:
    print("Error parsing Excel:", e)

File: dump_excel.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// Pure JS ZIP Reader
function readZipEntries(buffer) {
  const entries = {};
  let offset = 0;

  while (offset < buffer.length - 30) {
    const signature = buffer.readUInt32LE(offset);
    if (signature !== 0x04034b50) break;

    const compression = buffer.readUInt16LE(offset + 8);
    const compressedSize = buffer.readUInt32LE(offset + 18);
    const uncompressedSize = buffer.readUInt32LE(offset + 22);
    const nameLen = buffer.readUInt16LE(offset + 26);
    const extraLen = buffer.readUInt16LE(offset + 28);

    const name = buffer.toString('utf8', offset + 30, offset + 30 + nameLen);
    const dataStart = offset + 30 + nameLen + extraLen;

    if (compressedSize > 0 && dataStart + compressedSize <= buffer.length) {
      const compData = buffer.slice(dataStart, dataStart + compressedSize);
      let data;
      if (compression === 8) {
        try { data = zlib.inflateRawSync(compData); } catch (e) {
          try { data = zlib.inflateSync(compData); } catch (e2) {}
        }
      } else if (compression === 0) {
        data = compData;
      }
      if (data) {
        entries[name] = data.toString('utf8');
      }
    }

    offset = dataStart + (compressedSize || uncompressedSize);
    while (offset < buffer.length - 4 && buffer.readUInt32LE(offset) !== 0x04034b50) {
      offset++;
    }
  }
  return entries;
}

try {
  const excelPath = path.join(__dirname, 'SEO Files', 'issues_overview_report.xlsx');
  const xlsxBuf = fs.readFileSync(excelPath);
  const entries = readZipEntries(xlsxBuf);

  // 1. Parse Shared Strings
  const sharedStrings = [];
  const ssXml = entries['xl/sharedStrings.xml'];
  if (ssXml) {
    const regex = /<t[^>]*>([\s\S]*?)<\/t>/g;
    let match;
    while ((match = regex.exec(ssXml)) !== null) {
      sharedStrings.push(match[1]);
    }
  }

  // 2. Parse Sheet dynamically
  const sheetKey = Object.keys(entries).find(k => k.startsWith('xl/worksheets/'));
  if (!sheetKey) {
    console.error("No worksheet xml found in zip! Available files:", Object.keys(entries));
    process.exit(1);
  }
  const sheetXml = entries[sheetKey];

  // Find all rows
  const rowRegex = /<row[^>]*>([\s\S]*?)<\/row>/g;
  
  const rows = [];
  let rowMatch;
  while ((rowMatch = rowRegex.exec(sheetXml)) !== null) {
    const rowContent = rowMatch[1];
    const rowNum = parseInt(rowMatch[0].match(/r="(\d+)"/)[1]);
    const rowCells = {};
    
    // Regex for both string and numeric/inline cells
    const cellParser = /<c r="([A-Z]+)\d+"([^>]*)>(?:<v>([^<]+)<\/v>)?/g;
    let cellMatch;
    while ((cellMatch = cellParser.exec(rowContent)) !== null) {
      const colLetter = cellMatch[1];
      const attrs = cellMatch[2];
      const val = cellMatch[3];
      
      let finalVal = "";
      if (val !== undefined) {
        if (attrs.includes('t="s"')) {
          const strIdx = parseInt(val);
          finalVal = sharedStrings[strIdx] || "";
        } else {
          finalVal = val;
        }
      }
      rowCells[colLetter] = finalVal;
    }
    rows.push({ rowNum, cells: rowCells });
  }

  // Build Markdown report
  let md = "# Screaming Frog Excel Audit — Full Issue Details\n\n";
  
  md += "| Issue Name | Issue Type | Priority | URLs | % of Total | Description | How To Fix | Help URL |\n";
  md += "|---|---|---|---|---|---|---|---|\n";

  for (let i = 1; i < rows.length; i++) {
    const c = rows[i].cells;
    const name = c['A'] || "";
    const type = c['B'] || "";
    const priority = c['C'] || "";
    const urls = c['D'] || "";
    const pct = c['E'] || "";
    const desc = c['F'] || "";
    const fix = c['G'] || "";
    const help = c['H'] || "N/A";

    md += `| **${name}** | ${type} | ${priority} | ${urls} | ${pct}% | ${desc} | ${fix} | ${help} |\n`;
  }

  fs.writeFileSync(path.join(__dirname, 'SEO Files', 'excel_issues_dump.md'), md);
  console.log("SUCCESS: Generated excel_issues_dump.md");
} catch (e) {
  console.error("Error parsing Excel:", e);
}
